pkg/controllers/management/cis/scheduled_scan.go

```diff
@@ -0,0 +1,266 @@
+package cis
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/rancher/rancher/pkg/controllers/user/cis"
+	managementv3 "github.com/rancher/types/apis/management.cattle.io/v3"
+	v3 "github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/robfig/cron"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	DefaultCronSchedule    = "*/3 * * * *"
+	DefaultRetention       = 3
+	ScheduledScanKeyPrefix = "scheduledscan/"
+	separator              = ":"
+)
+
+type scheduleInfo struct {
+	CronSchedule       string
+	lastScanLaunchedAt time.Time
+	nextScanAt         time.Time
+}
+
+type scheduledScanHandler struct {
+	managementClient  managementv3.Interface
+	clusterClient v3.ClusterInterface
+	clusterLister     v3.ClusterLister
+	clusterScanClient v3.ClusterScanInterface
+	clusterScanLister v3.ClusterScanLister
+	clustersMap       map[string]*scheduleInfo
+}
+
+func NewScheduleScanHandler(
+	managementClient managementv3.Interface,
+	clusterClient v3.ClusterInterface,
+	clusterLister v3.ClusterLister,
+	clusterScanClient v3.ClusterScanInterface,
+	clusterScanLister v3.ClusterScanLister,
+) *scheduledScanHandler {
+	return &scheduledScanHandler{
+		managementClient:  managementClient,
+		clusterClient: clusterClient,
+		clusterLister:     clusterLister,
+		clusterScanClient: clusterScanClient,
+		clusterScanLister: clusterScanLister,
+		clustersMap:       make(map[string]*scheduleInfo),
+	}
+}
+
+func (i *scheduleInfo) String() string {
+	return fmt.Sprintf("CronSchedule: %v lastAt: %v nextAt: %v",
+		i.CronSchedule, i.lastScanLaunchedAt.String(), i.nextScanAt.String())
+}
+
+func (ssh *scheduledScanHandler) clusterSync(key string, cluster *v3.Cluster) (runtime.Object, error) {
+	if cluster == nil ||
+		cluster.DeletionTimestamp != nil ||
+		cluster.Spec.ScheduledClusterScan == nil {
+		return nil, nil
+	}
+
+	clusterInfo, ok := ssh.clustersMap[cluster.Name]
+	if cluster.Spec.ScheduledClusterScan.Enabled {
+		if !ok {
+			// Do stuff to enable scheduling
+			cronSchedule := getCronSchedule(cluster)
+			clusterInfo = &scheduleInfo{
+				CronSchedule: cronSchedule,
+			}
+			logrus.Infof("scheduledScanHandler: clusterSync: Scheduled Scan enabled for cluster %v: %v",
+				cluster.Name, cronSchedule)
+			if err := scheduleScan(cluster, ssh.clusterScanClient, clusterInfo); err != nil {
+				return nil, fmt.Errorf("scheduledScanHandler: clusterSync: error scheduling scan: %v", err)
+			}
+			ssh.clustersMap[cluster.Name] = clusterInfo
```

Comment:
multiple clusters could access the same map, we'd need to handle that using lock or syncMap

Reply:
will do.